src/platform_impl/windows/event_loop.rs

```diff
@@ -312,8 +325,11 @@ impl EventLoop {
             }
         }
 
-        self.wait_and_dispatch_message(timeout);
-
+        if self.exit_code().is_none() {
+            self.wait_for_messages(timeout);
+        }
```

Comment:
I have some concern here. Right now, it looks like this sequence happens here:
1. Call to `runner.wakeup()`.
2. Call to `wait_for_messages`
3. Inside `wait_for_messages` immeadiately calls `runner.prepare_wait()`.
4. Waits for events.
5. Call `runner.wakeup()`
6. `wait_for_messages` finishes
7. dispatch events
8. Call `runner.prepare_wait()`

My concern that there is no check if there are any new events between first `wakeup` and first `prepare_wait`. Maybe we should do sequence of `wakeup`. dispatch, `wait_for_messages`, dispatch, `prepare_wait`?

@rib I tag you because you previously asked to change calls to `wakeup` and `prepare_wait`.

Reply:
Hi, sorry that I don't have much time to follow the details here.

If we say that `prepare_wait` and `wakeup` are conceptually supposed to surround the "OS wait point" then it looks like we currently have two competing definitions of what that OS wait point should be...

It looks `pump_app_events` is intending to define the relinquish of control to the application loop (via `PumpStatus::Continue`) as the point where the Winit event loop will wait.

This arguably makes sense because `pump_app_events` is supposed to be non-blocking and any application using `pump_app_events` in their own loop should do something to throttle that loop before repeated called to `pump_app_events` (so applications using `pump_app_events` are arguably responsible for the wait, and they would also be expected to call `pump_app_events` after they wake up).

Sorry if I confused matters by overlooking these details when first commenting about changes here.

It looks like this needs to pick one definition.

Either:
1. Define the wait point by the act of relinquishing control to the app via `::Continue` and have some argument for `wait_for_messages()` that tells that function to skip over `prepare_wait` and `wakeup` when used from here.
2. Define the wait point as the call to `wait_for_messages_impl` and instead remove the call to `.wakeup()` at the start of `pump_app_events` and remove the call to `prepare_wait` just before we return with a `::Continue` status.

I guess option 1 would be most consistent with the current semantics.

Reply:
I think, maybe we should fix it in another PR? This PR doesn't change behaviour at this point and it is focused on fixing waiting precision.